### ISSUE TITLE
Should we switch to a permission-based authorisation model?

### ISSUE THREAD

**Issue by asimonw** [emfoundation/ce100-app#935]
Given the time it is taking to copy the admin role to the content owner role, I was wondering if there isn't a less brittle way to handle authorisations. If any page or feature only needs to check something like `user.hasPermission(page, feature)` and then the permissions are handled somewhere in one place in a config object, maybe that would be a good setup to add/update permissions and roles in the future?

That's of course one, possibly naive, solution, so this is rather a question. We will very likely need more complicated access schemes in the near future, so maybe worth thinking about.

**Comment by SimonLab** [comment]
I'm adding some initial thought on how the permission system currently works and how we could improve it:

A first filtering is done on when we define the routes. This is implemented by using the "scope" feature directly provided by Hapi (see https://blog.andyet.com/2015/06/16/harnessing-hapi-scopes/ for a tutorial). This first filter step will only allow users to access the endpoint if they have the necessary scope: "admin", "content-owner", "primary", "secondary". From the technical point of view to define this scope limitation:
- in the route folder (https://github.com/emfoundation/ce100-app/tree/master/server/routes) you can access all the files that define the endpoints (ex: /challenges, /organistions,...)
- If we look at the organisations endpoint file we can see that some routes define a scope:
https://github.com/emfoundation/ce100-app/blob/a7a0ab059474902e84275e1ed662128787a497b1/server/routes/orgs.js#L37
So the first step to allow or block users is to add or remove the scope in the config routes. I think this part is already well define and clear to understand.

The issue is when the permission rules are a bit more complex.

**Comment by iteles** [comment]
In light of #981, we need to carry out additional research _after_ #981 to determine whether our proposal above is still valid and determine whether it would be best to implement #981 ahead of this issue.

TBC whether this spike happens as part of Sprint 9, but the implementation will definitely not happen in this sprint.

**Comment by asimonw** [comment]
As agreed, we will limit this to a one-day spike for now to assess and implement potential refactoring/abstraction in order to make user permissions more manageable in the near future and possibly simplify upgrading to Hapi v17.

**Comment by iteles** [comment]
:+1: We had already carried out the initial estimate, adding the one day as discussed.

**Comment by iteles** [comment]
2 more days required to complete the implementation of our first iteration of this (including tests and documentation). Once #1043 has been estimated, @asimonw will determine whether the completion of _this_ issue fits into sprint 10 or should be postponed to sprint 11.

**Comment by asimonw** [comment]
@SimonLab Has this now been deployed as well?

**Comment by SimonLab** [comment]
@asimonw yes, will continue to add the permissions outside the handlers for the new endpoints and routes that we will add.

The permissions are now define in the configuration of the routes, for example:
https://github.com/emfoundation/ce100-app/blob/6d404ca4ce0ba32c111644dc5afd119d34ce2601/server/routes/people.js#L32-L34
The ```pre``` property means that the permissions will be checked before the real logic of the handler is run: https://hapijs.com/api#-routeoptionspre


"The pre option allows defining methods for performing actions before the handler is called. These methods allow breaking the handler logic into smaller, reusable components that can be shared ascross routes, as well as provide a cleaner error handling of prerequisite operations (e.g. load required reference data from a database)."

**Comment by iteles** [comment]
@SimonLab Let's make sure we add a PR with documentation on this before we close this issue please ✨

**Comment by asimonw** [comment]
@SimonLab Is it ok to close this one?

**Comment by SimonLab** [comment]
@asimonw I'm going to add some technical documentation for the permissions and then I think it should be ready for review/close
